src/TraceEvent/EventPipe/NetTraceFormat_v6.md

```diff
@@ -0,0 +1,288 @@
+
+In V5 an object header consisted of:
+- BeginPrivateObject Tag
+- SERIALIZED TYPE
+  - BeginPrivateObject Tag
+  - NullReference Tag
+  - version (int)
+  - minimum reader version (int)
+  - type name (string)
+  - EndPrivateObject Tag
+
+This header is 16 bytes + a variable length UTF8 string indicating the actual type of the object. The length of the object content was not included in the header. The EventBlock, MetadataBlock, and SPBlock objects included a BlockSize field as the first 4 bytes of the object content and the Trace object had fixed size.
+
+In V6 the object header is replaced by a simplified block header:
+
+```
+struct BlockHeader
+{
+  uint24 BlockSize
+  uint8 BlockKind
+}
+```
+
+Readers may want to read the BlockHeader as a single little endian 4 byte integer X where BlockSize = X & 0xFFFFFF and BlockKind = X >> 24.
+
+
+The type name strings used in the object header type name field map to new BlockKind values:
+
+```
+enum BlockKind
+{
+  EndOfStream = 0 // this is a new block type described later
+  Trace = 1,
+  Event = 2,
+  Metadata = 3,
+  Stack = 4,
+  SequencePoint = 5
+}
+```
+
+In V6 the EventBlock, MetadataBlock, StackBlock and SPBlock no longer have the initial BlockSize and padding fields. The value previously stored in the BlockSize object field is now in the BlockHeader. The semantics of the BlockSize header field are very similar to before, it represents the number of bytes in the remainder of the block. Specifically if file offset X points to the start of a block, then X + sizeof(BlockHeader) + BlockSize points to the start of the next block.
+
+Readers should skip over any blocks with an unrecognized BlockKind. New blocks might be introduced in minor version updates to store new types of data and this is expected to be a back-compatible change because old readers skip over it.
+
+
+#### Removed block footer
+
+In V5 all objects ended with a trailing EndPrivateObject tag byte. In V6 this trailing byte serves no purpose and is removed.
+
+#### Changed end of stream marker
+
+In V5 the file ended with a NullReferenceTag byte. In V6 the file ends with a new EndOfStream block. This block consists only of a BlockHeader and has BlockSize=0.
+
+## Multi-process support
+
+### Motivation
+
+Up to V5 the format was always designed to trace a single process only. Now we'd like to encode multi-process traces as well. This requires including process id information in places that previously only included thread ids. We also need to adjust definitions of a few TraceBlock fields.
+
+### Changes
+
+#### New ThreadBlock
+
+A new top-level block, BlockHeader.Kind=6, contains a set of thread id/process id tuples referenced by index. This allows many events to refer the same (thread id, process id) without needing to repeatedly encode the same pair of potentially large IDs.
+
+The content of a thread block is:
+
+- Count - uint16
+- Concatenated sequence of Count entries, each of which is
+  - Index - varuint
```

Comment:
Since this index must be unique for all events emitted between SequencePoints and if we could include events from multiple processes inside the same SequencePoint range, then the creation and handling of this block must be done by the process responsible to merge all events coming in from all traced processes into the nettrace stream. Each EventPipe session can't handle this since then there will be collision in indexes between processes included in the same nettrace file.

Reply:
No plan right now to have EventPipe do any merging or generate a multi-process trace. I agree that if we did ever want a scenario that took multiple single-process traces and merged them together it would require the merge to do substantial work with the current format. In addition to thread indexes there would also be collisions on metadata ids and stack ids + all the events would have to spliced together with specific ordering to obey the format's timestamp ordering rules.